File: 0200-number-of-islands/0200-number-of-islands.java
class Solution {

    int dx[] = {-1,0,0,1};
    int dy[] = {0,-1,1,0};
    public void dfs(char[][] grid, boolean[][] visited, int n, int m, int i, int j){
        visited[i][j] = true;

        for(int k=0; k<4; k++){
            int newx = i + dx[k];
            int newy = j + dy[k];
            if(newx >=0 && newy >=0 && newx < n && newy < m && !visited[newx][newy] && grid[newx][newy] == '1'){
                dfs(grid, visited, n, m, newx, newy);
            }

        }
    }
    public int numIslands(char[][] grid) {
        int n = grid.length;
        int m = grid[0].length;
        boolean[][] visited = new boolean[n][m];
        int count = 0;
        for(int i = 0; i< n ;i++){
            for(int j = 0 ; j<m ; j++){
                if(grid[i][j] == '1' && !visited[i][j]){
                dfs(grid, visited, n, m, i, j);
                count++;
                }
            }
        }
        return count;
    }
}

File: 0200-number-of-islands/0200-number-of-islands.cpp
class Solution {
void dfs(vector<vector<char>>& grid, int i, int j, int n, int m){
    if(i<0 || i>=n||j<0||j>=m || grid[i][j]=='0' ){
        return;
    }
    grid[i][j]='0';
    dfs(grid,i, j-1,n,m);
    dfs(grid,i, j+1,n,m);
    dfs(grid,i-1, j,n,m);
    dfs(grid,i+1, j,n,m);
}
public:
    int numIslands(vector<vector<char>>& grid) {
        int count=0;
        int n = grid.size();
        int m = grid[0].size();
        for(int i=0;i<n;i++){
            for(int j=0;j<m;j++){
                if(grid[i][j]=='1'){
                    dfs(grid,i, j,n,m);
                    count++;
                }
            }
        }
        return count;
    }
};